From a91265e3868278c75177fb8c447cf28dc5d56f70 Mon Sep 17 00:00:00 2001
From: Kent Tamura <tkent@chromium.org>
Date: Tue, 28 May 2024 05:15:11 +0000
Subject: [PATCH] RubyLB: Avoid crashes by `float` property

`float` in <ruby> has never worked correctly. This CL disables it by
updating `float` property in ComputedStyle, and adds a console message.

Bug: 324111880, 339098899
Change-Id: I0c537752a9b552b675690c3ce89f7488302aeaf3
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/5568921
Reviewed-by: Koji Ishii <kojii@chromium.org>
Auto-Submit: Kent Tamura <tkent@chromium.org>
Commit-Queue: Koji Ishii <kojii@chromium.org>
Cr-Commit-Position: refs/heads/main@{#1306598}
---

--- a/third_party/blink/renderer/core/css/resolver/style_adjuster.cc
+++ b/third_party/blink/renderer/core/css/resolver/style_adjuster.cc
@@ -80,6 +80,7 @@
 #include "third_party/blink/renderer/core/html/shadow/shadow_element_names.h"
 #include "third_party/blink/renderer/core/html_names.h"
 #include "third_party/blink/renderer/core/input_type_names.h"
+#include "third_party/blink/renderer/core/inspector/console_message.h"
 #include "third_party/blink/renderer/core/layout/layout_text_combine.h"
 #include "third_party/blink/renderer/core/layout/layout_theme.h"
 #include "third_party/blink/renderer/core/layout/list/list_marker.h"
@@ -640,10 +641,22 @@
   // We need to avoid to inlinify children of a <fieldset>, which creates a
   // dedicated LayoutObject and it assumes only block children.
   if (layout_parent_style.InlinifiesChildren() &&
-      !builder.HasOutOfFlowPosition() && !builder.IsFloating() &&
+      !builder.HasOutOfFlowPosition() &&
       !(element && IsA<HTMLFieldSetElement>(element->parentNode()))) {
-    builder.SetIsInInlinifyingDisplay();
-    builder.SetDisplay(EquivalentInlineDisplay(builder.Display()));
+    if (RuntimeEnabledFeatures::RubyLineBreakableEnabled() &&
+        builder.IsFloating()) {
+      builder.SetFloating(EFloat::kNone);
+      element->GetDocument().AddConsoleMessage(
+          MakeGarbageCollected<ConsoleMessage>(
+              ConsoleMessage::Source::kRendering, ConsoleMessage::Level::kInfo,
+              "`float` property is not supported correctly inside an element "
+              "with `display: ruby` or `display: ruby-text`."),
+          true);
+    }
+    if (!builder.IsFloating()) {
+      builder.SetIsInInlinifyingDisplay();
+      builder.SetDisplay(EquivalentInlineDisplay(builder.Display()));
+    }
   }
 
   if (builder.Display() == EDisplay::kBlock) {
